opentelemetry-sdk/src/metrics/internal/mod.rs

```diff
@@ -164,35 +170,40 @@ where
     where
         MapFn: FnMut(Vec<KeyValue>, A) -> Res,
     {
-        prepare_data(dest, self.count.load(Ordering::SeqCst));
-        if self.has_no_attribute_value.swap(false, Ordering::AcqRel) {
+        // reset sorted trackers so new attributes set will be written into new hashmap
+        let trackers = match self.sorted_attribs.lock() {
```

Comment:
I believe this is making the contention between collect and update worse. Earlier, the only point of contention during collect was the write lock under which we replace the existing `RwLock<Hashmap>` with a default one. With the changes in this PR, we now have two points of contention during collect:
1. Locking the `Mutex` with `self.sorted_attribs.lock()` which would compete with updates trying to add newer attribute sets
2. Acquiring the write lock with `self.all_attribs.write()` which would compete with all the updates

Reply:
Yes, I know, but I don't see other way if we want to implement sharding, so my goal with this PR was mostly try to preserve existing performance (as much as I could) while at the same time, make it easy to add sharding in the future.
However, I expect this extra lock to be very cheap in practice, for two reasons:
1) Mutex is generally fast
2) at the end of collection cycle, there should be quite low contention on inserting new attribute sets.

Reply:
> but I don't see other way if we want to implement sharding, 

Why is that? Implementing sharding for the Hashmap should be only about replacing the existing `HashMap` type with the type that implements sharding something like `Dashmap`. Why do we need two sets of attributes trackers for using a sharded Hashmap?

Reply:
Everything could work with one hashmap, we could probably tolerate slower cumulative collection phase, but the real problem is with delta temporality collection.
Here's the problem:
We have hard requirement that attribute-set keys are order agnostic. 
In practice this means, that we can have multiple attribute-sets in hashmap, which points to same tracker (hence, we have `Arc<A>`.).
With sharding this means, that same tracker can be in multiple shards, so the question is how do you reset hashmap?
* we cannot simply "replace" hashmap with fresh instance.
* we cannot iterate through each shard, deduplicate (using hashset to identify already processed attribute-sets), and reset/clear each shard while iterating. This doesn't work because, even if you remove attribute-set in one shard, on next tracker update same tracker instance will be put back, as long as sorted attribute-set instance lives in another shard.

The only option to actually clear sharded hashmap is to lock ALL shards, then iterate/clone and unlock.

So my approach is to have two hashmaps instead, because all things considered it has more benefits than drawbacks.
* update existing tracker - no effect, because attribute set is found on first try in both cases (in case user never provides sorted attribute-sets, it might be even better, as first hashmap will be half as small, so less chance for collision).
* inserting new tracker - probably not expected, but it is performance increase actually (with cargo bench --bench metrics_counter), probably due to less checks and fast lock with Mutex.
* collection phase (both delta and cumulative) - huge benefits, short lock time, no need to have hashset in order to dedup, and solves #2093 (which would further slow down collection phase).

So basically this approach has bunch of benefits and 0 drawbacks, so why not try it :)

Reply:
I understand that sharding has proven to be improving throughout significantly, but those tests were done at the time we took a Mutex lock in hotpath. Now we have replaced that with Read lock on hotpath, so there is no contention anymore. i.e if 10 threads wants to update(), they can all do so without being blocked by each other. With this existing design, how does sharding help improve performance?